amethyst_derive/src/prefab_data.rs

```diff
@@ -33,47 +36,216 @@ fn impl_prefab_data_component(ast: &DeriveInput) -> TokenStream {
+                }),
+                None => {
+                    let var_name =
+                        Ident::new(&format!("field_{}", field_number), Span::call_site());
+                    let number = Literal::usize_unsuffixed(field_number);
+                    Some(quote! {
+                        let #var_name = &mut self.#number;
+                    })
+                }
             }
+        } else {
+            None
         }
     });
-    let adds = (0..data.len()).map(|n| {
-        let tuple_index = Literal::usize_unsuffixed(n);
-        let (_, name, is_component) = &data[n];
+    (
+        data_types,
+        quote! {
+            #(#extract_fields_add)*
+            #(#add_to_entity)*
+        },
+        quote! {
+            #(#extract_fields_sub)*
+            #(#subs)*
+        },
+    )
+}
+
+fn prepare_prefab_aggregate_enum(
+    base: &Ident,
+    data: &DataEnum,
+) -> (Vec<(Type, bool)>, TokenStream, TokenStream) {
+    let mut data_types = Vec::new();
+    let mut subs = Vec::new();
+    let mut add_to_entity = Vec::new();
+
+    for variant in &data.variants {
+        let (variant_add_to_entity, variant_subs) =
+            prepare_prefab_aggregate_fields(&mut data_types, &variant.fields);
+        let field_names_add: Vec<_> = variant
+            .fields
+            .iter()
+            .enumerate()
+            .map(|(field_number, field)| match &field.ident {
+                Some(name) => name.clone(),
+                None => Ident::new(&format!("field_{}", field_number), Span::call_site()),
+            })
+            .collect();
+        let field_names_sub: Vec<_> = variant
+            .fields
+            .iter()
+            .enumerate()
+            .map(|(field_number, field)| match &field.ident {
+                Some(name) => {
+                    if !have_component_attribute(&field.attrs[..]) {
+                        quote! {
+                            #name
+                        }
+                    } else {
+                        quote! {
+                            #name: _
+                        }
+                    }
+                }
+                None => {
+                    let var_name = if !have_component_attribute(&field.attrs[..]) {
+                        Ident::new(&format!("field_{}", field_number), Span::call_site())
+                    } else {
+                        Ident::new(&format!("_field_{}", field_number), Span::call_site())
+                    };
+                    quote! {
+                        #var_name
+                    }
+                }
+            })
+            .collect();
+        let ident = &variant.ident;
+        add_to_entity.push(match variant.fields {
+            Fields::Named(_) => quote! {
+                #base::#ident {#(#field_names_add,)*} => {
+                    #(#variant_add_to_entity)*
+                }
+            },
+            Fields::Unnamed(_) => quote! {
```

Comment:
does it work if you collapse this with the previous branch? like

```rust
Fields::Named(_) | Fields::Unnamed(_) =>
```

same for `subs`

Reply:
Sadly, it doesn't. I'm producing tuple matching syntax in this branch and struct matching syntax in the other. I think I could remove some of the duplication from the two `quote!` blocks but I think at the cost of clarity.